Add `--stat` option

alex-saunders opened the pull request:
- Adds `--stat` option to CLI
- When `--stat` is passed:
  - Direct + shared dependencies split and their sizes printed to console
  - Any difference in dependency size (from the last time `snowpack` was ran) is printed
- Functionality provided by new rollup plugin: `rollup-plugin-dependency-info` 

- Current output format copied from https://github.com/pikapkg/snowpack/issues/163#issuecomment-578448785 but I've tried to make the output easily changable 
  - Output doesn't use `chalk` yet as I wasn't sure what colors/styles would be desirable).

Sample output:
<img width="584" alt="Screenshot 2020-02-05 at 13 59 45" src="https://user-images.githubusercontent.com/22820481/73848326-ce447380-481f-11ea-92f0-3fa9010b7bb2.png">

Closes #163

alex-saunders commented:
The `node_modules` diff is a bit crazy so not sure if I've done something wrong with adding my new test?

Also the tests are failing because it looks like the installed dependency sizes differs on Windows + Linux, not totally sure how to get around this without updating the test conditions to ignore the precise number printed in the stat output.

FredKSchott commented:
ah yea, it looks like the dependencies you chose actually have quiet a few sub-dependencies. You might want to replace them with others. Take a look at https://www.pika.dev/search?q= and try to find some packages with the "0" icon, indicating zero sub-deps. `vuex`, `acorn`, and `path-to-regexp` may be better alternatives to `react-router`,  `react-redux`, and `formik`.

regarding the failing tests: In our test runner, we already have some logic for stripping the benchmark (ex: `[0.12s]`) out of our test output. You could probably write something similar, or just get rid of the `s` in `stripBenchmark()` so that our tests ignore anything inside of square brackets when they compare. 

```diff
function stripBenchmark(stdout) {
-  return stdout.replace(/\s*\[\d+\.?\d+s\](\n?)/g, '$1'); //remove benchmark
+  return stdout.replace(/\s*\[.*?\](\n?)/g, '$1'); //remove any benchmarking data
}

FredKSchott commented:
this looks great! Really excited to get this in!

The diff is so large that I can't load the  src/ folder in GH to review. if you can shrink the number of deps in the test to something more manageable then I can review.

stramel commented:
> Output doesn't use chalk yet as I wasn't sure what colors/styles would be desirable).

Some quick ideas:
- Have the diff show red for increase and green for decrease in size?
- Have the size show red, yellow, green based on thresholds similarly to how the Pika website does?
- Have the package name colored a blue/teal?

Ultimately, coloring will be pretty opinionated and I will defer to @FredKSchott's preferences.

alex-saunders commented:
> - Have the diff show red for increase and green for decrease in size?
> - Have the package name colored a blue/teal? or green to match the above list of packages?

I was thinking pretty much the same as this so if @FredKSchott is happy then this coloring sounds good.

> - Have the size show red, yellow, green based on thresholds similarly to how the Pika website does?

I love this idea! @FredKSchott do you know what exact thresholds the pika website is using? I love the idea of copying the design already present in the pika-verse.

alex-saunders commented:
@FredKSchott I've managed to find 2 packages that both rely on a single dependency which makes for a pretty nice, minimal test case. The PR should now be much more manageable to review so feel free to have a look while I work on fixing the test cases & colouring the output!

FredKSchott commented:
lol from 2000+ new files to 46. Nice work 👍👍👍

alex-saunders commented:
I've got a strange failing test - looking like the actual runner is failing due to some output from `dircompare` potentially being undefined?

FredKSchott commented:
> > * Have the size show red, yellow, green based on thresholds similarly to how the Pika website does?
> 
> I love this idea! @FredKSchott do you know what exact thresholds the pika website is using? I love the idea of copying the design already present in the pika-verse.

Yes! here's the code I use on pika.dev

```
    if (stat < 10) {
      statType = 'is-success';
    } else if (stat < 25) {
      statType = 'is-warning';
    } else {
      statType = 'is-danger';
    }
```

However, that's gzipped & minified. You may want to use something closer my original suggestion from the convo that led to this feature: https://twitter.com/FredKSchott/status/1075574385069576193

Also, I wonder if color scale should change if you do `--optimize` vs. not optimized, since that changes your sizes. I think actually not, since keeping a single scale for both would highlight why `--optimize` is better for production, if you care about stats. Maybe instead of changing the color scale, we add a small note/warning if you use `--stat` without `--optimize`.

FredKSchott commented:
> I've got a strange failing test - looking like the actual runner is failing due to some output from `dircompare` potentially being undefined?
> Annoyingly it's only on Windows and I don't have a windows machine so a little hard for me to debug this

Hmm, can you add a console.log() to the test runner? Something like: 

```diff
res.diffSet.forEach(function(entry) {
+  console.log(entry);
```

It looks like either path2 or entry2 is undefined, but I'm curious why / what could be going on.

FredKSchott commented on src/index.ts:
```diff
@@ -45,6 +47,7 @@ const ALWAYS_SHOW_ERRORS = new Set(['react', 'react-dom']);
 const cwd = process.cwd();
 const banner = chalk.bold(`snowpack`) + ` installing... `;
 const installResults = [];
+let dependencyInfo: DependencyInfoOutput = null;
```
nit: is `DependencyInfoOutput` output, or output info? May be better named as `DependencyOutputInfo` or `DependencyOutputStats`.

FredKSchott commented on src/index.ts:
```diff
@@ -80,6 +83,28 @@ function formatInstallResults(skipFailures): string {
 
+function formatDependencyInfoAnalysis(): string {
+  let output = '';
+  const {direct, shared} = dependencyInfo;
```
In a future PR, I'd love to move this variable out of the global context and pass it in as a function argument instead. But, since we have a few other global variables like this one already, this is totally fine to leave as-is for now.

alex-saunders replied:
Agreed, I only made this global to copy the currently existing pattern. I'm happy to move this into an argument now, however, it should be pretty easy to do so?

FredKSchott commented on src/index.ts:
```diff
@@ -80,6 +83,28 @@ function formatInstallResults(skipFailures): string {
+function formatDependencyInfoAnalysis(): string {
+  let output = '';
+  const {direct, shared} = dependencyInfo;
+  const formatFileInfo = files =>
```
nit:

```diff
function formatFileInfo(file: NEEDS_TYPE) { ... }

- output += formatFileInfo(Object.values(direct)).join('\n');
+ output += Object.values(direct).map(formatFileInfo).join('\n');
```

FredKSchott commented on src/index.ts:
```diff
@@ -80,6 +83,28 @@ function formatInstallResults(skipFailures): string {
+  const formatFileInfo = files =>
+    files.map((file, index) => {
+      const {fileName, size, delta} = file;
+      let fileInfo = `${index >= files.length - 1 ? '└─' : '├─'} ${fileName} [${Math.round(
```
nit: it would be great to move these variables out of the string declaration:

```
const lineGlyph = index >= files.length - 1 ? '└─' : '├─';
const lineSize = Math.round((size / 1000) * 100) / 100 + ' KB';
const lineDelta = // ...
return `${lineGlyph} ${fileName} [${lineSize}${lineDelta}]`;
```

FredKSchott commented on src/rollup-plugin-dependency-info.ts:
```diff
@@ -0,0 +1,68 @@
+
+export function rollupPluginDependencyInfo(cb: (dependencyInfo: DependencyInfoOutput) => void) {
+  let outputDir: string;
+  let cache: {
```
nit: this should be able to go on one line. But, no worries if prettier won't let it.

FredKSchott commented on src/rollup-plugin-dependency-info.ts:
```diff
@@ -0,0 +1,68 @@
+  } = {};
+  let output: DependencyInfoOutput = {
+    direct: {},
+    shared: {},
```
nit: can we use `common` here instead of `shared`? I'd love to keep the naming variations to a minimum, and since our rollup config is already using `web_modules/common` it would be nice to stick to that.

FredKSchott commented on src/rollup-plugin-dependency-info.ts:
```diff
@@ -0,0 +1,68 @@
+    generateBundle(options: OutputOptions, bundle: OutputBundle) {
+      outputDir = options.dir;
+
+      for (let fileName of Object.keys(bundle)) {
```
It would help readers to understand what's going on if you pull this out into its own `buildCache()` function, above `compareDependencies()`.

FredKSchott commented on src/rollup-plugin-dependency-info.ts:
```diff
@@ -0,0 +1,68 @@
+    writeBundle(bundle: OutputBundle) {
+      const files = Object.keys(bundle);
+
+      const [directDependencies, sharedDependencies] = files.reduce(
```
It's a bit hard to understand what's going on here. Can you replace `d`/`s`/`f` with more descriptive names, or add a comment above this line to explain whats going on?

FredKSchott commented on src/rollup-plugin-dependency-info.ts:
```diff
@@ -0,0 +1,68 @@
+    shared: {},
+  };
+
+  const compareDependencies = (files: string[], type: DependencyType) => {
```
nit: `function compareDependencies(files: string[], type: DependencyType) {`

FredKSchott commented on src/rollup-plugin-dependency-info.ts:
```diff
@@ -0,0 +1,68 @@
+import path from 'path';
+import {OutputOptions, OutputBundle} from 'rollup';
+
+type Dependency = {
```
Since `Dependency` / `Dependencies` / `DependencyType` are all data types that get exposed/exported to the rest of the program through `DependencyInfoOutput`, we need to give them less generic names so that we don't get confused.

`DependencyStats`, `DependencyStatsMap`, `DependencyStatsType`?

alex-saunders commented:
@FredKSchott Thanks so much for the detailed review, I think I've addressed all of your comments, let me know if I missed anything or you want anything changed. It's been a lovely codebase to work in so I'm happy to change things if my style doesn't fit in with what's currently present.

I've also had a go at coloring the output, given the discussion above. 
I have the file names in cyan, the file sizes follow this logic:
```
let color;
if (kb < 15) {
  color = 'green';
} else if (kb < 30) {
  color = 'yellow';
} else {
  color = 'red';
}
```
and the delta size appears in red or green dependent on whether it has grown or shrunk, respectively.

alex-saunders commented:
Looks like the tests are failing due to `dircompare` trying to compare the expected `common` installs but the hashes are generated differently on windows so it is complaining that `noop-x.esm-e72ec11e` can't be found in the generated files.. 
Is there any way to tell dircompare to ignore hashes in file names?

FredKSchott commented:
@alex-saunders good catch, lets just ignore it for now. can you try adding `excludeFilter: 'common'` to the ` dircompare.compareSync` call options (/w a comment explaining why)?

FredKSchott commented on __tests__/integration/runner.js:
```diff
@@ -8,12 +8,23 @@ const dircompare = require('dir-compare');
 function stripRev(code) {
   return code.replace(/\?rev=\w+/gm, '?rev=XXXXXXXXXX');
 }
+function stripChunkHashe(stdout) {
```
nit: `stripChunkHash`

FredKSchott commented on __tests__/integration/runner.js:
```diff
@@ -59,6 +73,7 @@ for (const testName of readdirSync(__dirname)) {
     });
     // If any diffs are detected, we'll assert the difference so that we get nice output.
     res.diffSet.forEach(function(entry) {
+      console.log(entry);
```
nit: remember to remove once you ignore common

alex-saunders commented:
@FredKSchott Looks like I've finally managed to get the tests passing 🎉so we're probably good to go?

FredKSchott commented:
Tested locally and it looks great! Thanks for all your work on this, if feedback is good we may even want to make this the default output!

stramel commented:
Do either of you have a screenshot of what it looks like with color?

FredKSchott commented:
I might make some small changes to the output, but here was a quick screenshot I took before starting:

<img width="402" alt="Screen Shot 2020-02-07 at 5 39 49 PM" src="https://user-images.githubusercontent.com/622227/74076977-ff17da80-49d0-11ea-880e-fe70fe84260c.png">

Here's what I'm playing around with now (will put into PR):
<img width="712" alt="Screen Shot 2020-02-07 at 5 41 20 PM" src="https://user-images.githubusercontent.com/622227/74077000-12c34100-49d1-11ea-906b-87f755abe2b6.png">
